File: packages/some-content/public/code-samples/two-sum.ts
/**
 * Two Sum
 *
 * Given an array of integers `nums` and an integer `target`, return the
 * indices of the two numbers that add up to `target`.
 *
 * Trades a second pass for O(1) lookups: a single scan builds up `seen`
 * while checking, at each index, whether this number's complement has
 * already been visited.
 */
export function twoSum(nums: number[], target: number): [number, number] {
  const seen = new Map<number, number>()

  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i]!
    const match = seen.get(complement)
    if (match !== undefined) {
      return [match, i]
    }
    seen.set(nums[i]!, i)
  }

  throw new Error("no two sum solution")
}

File: packages/some-content/public/code-samples/two-sum.rs
use std::collections::HashMap;

/// Two Sum
///
/// Given a slice of integers `nums` and a `target`, return the indices of
/// the two numbers that add up to `target`.
///
/// A single pass fills `seen` while checking, at each index, whether this
/// number's complement has already been visited - no second pass needed.
pub fn two_sum(nums: &[i32], target: i32) -> Option<(usize, usize)> {
    let mut seen: HashMap<i32, usize> = HashMap::new();

    for (i, &num) in nums.iter().enumerate() {
        let complement = target - num;
        if let Some(&j) = seen.get(&complement) {
            return Some((j, i));
        }
        seen.insert(num, i);
    }

    None
}
